src/main/java/dev/blaauwendraad/masker/json/KeyContainsMasker.java

```diff
@@ -164,38 +161,103 @@ public byte[] mask(byte[] input) {
+                && !keyMaskingConfig.isDisableBooleanMasking()) {
+            maskBooleanValueInPlace(maskingState, keyMaskingConfig);
+        } else {
+            return false;
+        }
+        return true;
+    }
+
     /**
      * Checks if the current byte changes the current json path. If so, updates current json path.
      */
     private void trackCurrentJsonPath(MaskingState maskingState) {
-        // Check if this is the end of the current json object
-        if (!maskingState.isInArray() && currentByteIsUnescapedCurlyBracketClose(maskingState)) {
+        // special case of empty json objects: look ahead of the curly brackets opening to check if the key is an empty
+        // json object. If so, put a placeholder onto the current json path stack. If no, rollback to the current index
+        if (isCurlyBracketOpen(maskingState.byteAtCurrentIndex()) && !maskingState.isInArraySegment()) {
+            int currentIndex = maskingState.currentIndex();
+            maskingState.incrementCurrentIndex();
+            skipWhitespaceCharacters(maskingState);
+            if (isCurlyBracketClose(maskingState.byteAtCurrentIndex())) {
+                maskingState.expandCurrentJsonPath(maskingState.byteAtCurrentIndex(), 1);
+            } else {
+                maskingState.setCurrentIndex(currentIndex);
+            }
+        }
+
+        // Check if this is the end of the current non-empty json object
+        if (currentByteIsUnescapedCurlyBracketClose(maskingState) && !maskingState.isInRootSegment() && !maskingState.isInArraySegment()) {
             maskingState.backtrackCurrentJsonPath();
         }
         // Check if this is the end of the current json array
-        if (maskingState.isInArray() && currentByteIsUnescapedSquareBracketClose(maskingState)) {
+        if (currentByteIsUnescapedSquareBracketClose(maskingState) && maskingState.isInArraySegment()) {
             maskingState.backtrackCurrentJsonPath();
-            // if it is not a nested array, backtrack the parent key component as well
-            if (!maskingState.isInArray()) {
-                maskingState.backtrackCurrentJsonPath();
-            }
-        }
-        // Check if this is the end of the current json array element
-        if (maskingState.isInArray() && currentByteIsUnescapedComma(maskingState)) {
-            maskingState.incrementCurrentJsonPathArrayIndex();
         }
         // Check if this is the start of a json array
         if (currentByteIsUnescapedSquareBracketOpen(maskingState)) {
             maskingState.expandCurrentJsonPath();
+            KeyMaskingConfig keyMaskingConfig = keyMatcher.getMaskConfigIfMatched(maskingState.getMessage(), maskingState.getCurrentJsonPath());
+            if (keyMaskingConfig != null) {
+                maskArrayElementInPlace(maskingState, keyMaskingConfig);
+                trackCurrentJsonPath(maskingState);
+            }
+        }
+        if (currentByteIsUnescapedComma(maskingState)) {
```

Comment:
Might be nice for clarity to add one line of JavaDoc explaining you are checking if there is a next element here.

Reply:
I will do it.